packages/common/src/core/nextOccuranceUtils.ts

```diff
@@ -0,0 +1,220 @@
+  } else {
+    // Set the event time to the next occurrence of the specified day at the given hour
+    targetTime.setDate(dayOfMonth);
+  }
+
+  targetTime.setHours(hour, 0, 0, 0);
+
+  // Convert the event time to an ISO string
+  return targetTime;
+};
+
+/**
+ * Get the next yearly occurance of the specified hour in the specified timezone
+ * @param hour takes an hour of the day between 0 and 23
+ * @param timezone takes a timezone string
+ * @param dayOfMonth takes a day of the week between 1 and 28
+ * @param month takes a month of the year between 0 and 11
+ * @returns a Date object representing the next yearly occurance of the specified hour
+ */
+const getNextYearlyOccurance = (
+  hour: number,
+  timezone: string,
+  dayOfMonth: number,
+  month: number
+): Date => {
+  validateHour(hour);
+  validateDayOfMonth(dayOfMonth);
+  validateMonth(month);
+
+  // Get the time in the specified timezone
+  const targetTime = getInitialTargetTime(timezone, "en-US");
+
+  // Calculate the next occurrence of the specified month and day
+  const currentMonth = targetTime.getMonth();
+  const currentDayOfMonth = targetTime.getDate();
+  const currentYear = targetTime.getFullYear();
+
+  if (
+    currentMonth > month ||
+    (currentMonth === month &&
+      (currentDayOfMonth > dayOfMonth ||
+        (currentDayOfMonth === dayOfMonth && targetTime.getHours() >= hour)))
+  ) {
+    // If the specified month and day have already passed this year, schedule it for next year
+    targetTime.setFullYear(currentYear + 1);
+  }
+  targetTime.setMonth(month);
+  targetTime.setDate(dayOfMonth);
+  targetTime.setHours(hour, 0, 0, 0);
+
+  // Convert the event time to an ISO string
+  return targetTime;
+};
+
+/**
+ * Get the next occurance of the specified schedule
+ * @param schedule takes an IHubSchedule object
+ * @returns a Date object representing the next occurance of the specified schedule
+ */
+export const getNextOccurance = (schedule: IHubSchedule): Date => {
+  if (schedule.mode === "scheduled") {
+    switch (schedule.cadence) {
+      case "daily":
+        return getNextDailyOccurance(schedule.hour, schedule.timezone);
+      case "weekly":
+        return getNextWeeklyOccurance(
+          schedule.hour,
+          schedule.timezone,
+          schedule.day
+        );
+      case "monthly":
+        return getNextMonthlyOccurance(
+          schedule.hour,
+          schedule.timezone,
+          schedule.date
+        );
+      case "yearly":
+        return getNextYearlyOccurance(
+          schedule.hour,
+          schedule.timezone,
+          schedule.date,
+          schedule.month
+        );
+      default:
+        throw new Error("Invalid cadence");
+    }
+  } else {
+    // If the schedule mode is not "scheduled", return undefined because it is either "automatic" or "manual"
+    return undefined;
+  }
+};
+
+/****** REPEATED CODE ******/
+const getInitialTargetTime = (timezone: string, locale: string): Date => {
```

Comment:
It's hard for me to follow what this function is doing and why all the date manipulation is necessary. Could you please add a comment to explain?